### PR TITLE
Kinesis Stream Data Storage

### PR DESCRIPTION
**Pull request by Dala0**
## What does this change?

Adds writing of data about requests and responses to a Kinesis stream.

## How to test

The new code should write data to the Kinesis stream called 'encodings_endpoint_data'.

## How can we measure success?

The correct data is written to the Kinesis stream.

### CODE REVIEW

File: endpoint/common.php [hunk 1]
```diff
@@ -7,21 +7,26 @@
 require $autoload_name;
 
 use Aws\Sns\SnsClient;
+use Aws\Kinesis\KinesisClient;
+use Aws\Exception\AwsException;
 
 /*stop AWS complaining*/
 date_default_timezone_set('UTC');
 
+$GLOBALS['kinesis_stream_name'] = "encodings_endpoint_data";
```
Comment (fredex42) [inline endpoint/common.php]: Would be better to set this externally, how are the other globals like `DB_HOST` set? My guess is that they come from the Cloudformation and the instance kickstart script echoes the values into `php.ini` or similar?

File: endpoint/common.php [hunk 2]
```diff
@@ -98,7 +103,8 @@ function find_content($config){
 			),
 			);
 		report_error($details);
-		header('HTTP/1.0 500 Server Config Error',true,500);
+    header('HTTP/1.0 500 Server Config Error',true,500);
+    write_to_kinesis($GLOBALS['kinesis_stream_name'], '{"access_url":"'.$GLOBALS['actual_link'].'", "output_message":null, "response_code":500, "php_headers":'.json_encode(headers_list()).'}', null);
```
Comment (fredex42) [inline endpoint/common.php:110]: surely php has nicer ways to serialize json than this?! Might as well leave it if it's working, I suppose, but there is always a risk that a dodgy character will slip through and break any reader

**Review by fredex42 — CHANGES_REQUESTED** [review]
Minor tweaks. When you deploy please check some of the interactives that use this to ensure that they are working.

File: endpoint/common.php (hunk 2 above)
Comment (Dala0) [reply endpoint/common.php:110]: Are you refering to my use of the json_encode() function or the use of strings to generate JSON?

File: endpoint/common.php (hunk 1 above)
Comment (Dala0) [reply endpoint/common.php]: Done.

File: endpoint/common.php (hunk 2 above)
Comment (fredex42) [reply endpoint/common.php:110]: strings to generate json! surely you could set up the whole thing as some kind of dictionary/object and use `json_encode` on that?

**Review by fredex42 — APPROVED** [review]
:+1:

File: endpoint/common.php (hunk 2 above)
Comment (Dala0) [reply endpoint/common.php:110]: That is possible yes. I am not sure how it would help.

Comment (fredex42) [reply endpoint/common.php:110]: Don't worry about it in this specific instance, but I would always do that in preference to manually coding strings. The reason is that then you should be guaranteed to be able to _parse_ it on the other side - you won't accidentally end up with unescaped characters breaking the parser.

File: endpoint/common.php [hunk 3]
```diff
@@ -480,9 +480,9 @@ function write_to_kinesis($name, $content, $output_message){
   }
 
   $kinesisClient = new Aws\Kinesis\KinesisClient([
-      'profile' => 'default',
       'version' => '2013-12-02',
-      'region' => 'eu-west-1'
+      'region' => 'eu-west-1',
```
Comment (fredex42) [inline endpoint/common.php:484]: By preference I would take this from the configuration file too, though I guess it does not matter too much in this instance.  The configuration file can then be set in the cloudformation via the ${AWS::Region} substitution which is automatically set to the region in which the cfn is being deployed

**Pull request merged**